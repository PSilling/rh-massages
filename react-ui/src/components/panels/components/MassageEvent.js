// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Tooltip } from "reactstrap";
import moment from "moment";

// component imports
import ConfirmationModal from "../../modals/ConfirmationModal";
import TooltipIconButton from "../../iconbuttons/TooltipIconButton";

// util imports
import _t from "../../../util/Translations";
import Auth from "../../../util/Auth";
import Util from "../../../util/Util";

/**
 * Custom component event for Massages calendar view.
 */
class MassageEvent extends Component {
  tooltipTarget = `MassageEventID${this.props.event.massage.id}`;

  constructor(props) {
    super(props);

    this.state = { modalActive: false, tooltipOpen: false };
  }

  componentWillUnmount() {
    this.setState({ modalActive: false, tooltipOpen: false });
  }

  assignWithCalendar = () => {
    window.open(Util.getEventLink(this.props.event.massage), "_blank");
    this.props.onAssign(this.props.event.massage);
  };

  toggleTooltip = () => {
    this.setState(prevState => ({ tooltipOpen: !prevState.tooltipOpen }));
  };

  toggleModal = () => {
    this.setState(prevState => ({
      modalActive: !prevState.modalActive,
      tooltipOpen: prevState.modalActive ? prevState.tooltipOpen : false
    }));
  };

  createIcons = () => {
    const icons = [];

    if (!Auth.isMasseur() && Util.isEmpty(this.props.event.massage.client)) {
      const assignDisabled =
        this.props.massageMinutes +
          moment(this.props.event.massage.ending).diff(moment(this.props.event.massage.date), "minutes") >
        Util.MAX_MASSAGE_MINS;
      icons.push(
        <TooltipIconButton
          key="assignCal"
          icon="calendar-plus"
          size="md"
          onClick={this.assignWithCalendar}
          disabled={assignDisabled}
        />
      );
      icons.push(
        <TooltipIconButton
          key="assign"
          icon="check"
          size="md"
          onClick={() => this.props.onAssign(this.props.event.massage)}
          disabled={assignDisabled}
        />
      );
    } else if (
      !Util.isEmpty(this.props.event.massage.client) &&
      Auth.getSub() === this.props.event.massage.client.sub
    ) {
      icons.push(
        <TooltipIconButton
          key="onlyCal"
          icon="calendar-plus"
          size="md"
          onClick={() => window.open(Util.getEventLink(this.props.event.massage), "_blank")}
        />
      );
      icons.push(
        <TooltipIconButton
          key="cancel"
          icon="times"
          size="md"
          onClick={() => this.props.onCancel(this.props.event.massage)}
        />
      );
    }

    if (Auth.isAdmin() || (Auth.isMasseur() && this.props.event.massage.masseuse.sub === Auth.getSub())) {
      icons.push(
        <TooltipIconButton
          key="edit"
          icon="edit"
          size="md"
          onClick={() => this.props.onEdit(this.props.event.massage)}
        />
      );
      icons.push(<TooltipIconButton key="delete" icon="trash" size="md" onClick={this.toggleModal} />);
    }

    icons.push(
      <TooltipIconButton
        key="more"
        icon="ellipsis-v"
        size="md"
        onClick={() => this.props.onShowMore(this.props.event)}
      />
    );

    return icons;
  };

  render() {
    return (
      <div id={this.tooltipTarget} style={{ minHeight: "100%", minWidth: "100%" }}>
        {this.props.view === "work_week" && (
          <span className="rbc-event-label" style={{ marginTop: "2px", marginBottom: "3px" }}>
            {`${moment(this.props.event.massage.date).format("LT")} â€” ${moment(this.props.event.massage.ending).format(
              "LT"
            )}`}
          </span>
        )}

        <span>
          {this.props.archived && `${this.props.event.massage.facility.name}: `}
          {this.props.event.massage.masseuse.name}
        </span>

        {this.props.activeTooltip === this.tooltipTarget && !this.props.archived && (
          <Tooltip
            isOpen={this.state.tooltipOpen}
            toggle={this.toggleTooltip}
            target={this.tooltipTarget}
            autohide={false}
          >
            <div style={{ marginLeft: "-2px", marginRight: "-2px" }}>{this.createIcons()}</div>
          </Tooltip>
        )}
        {this.state.modalActive && (
          <ConfirmationModal
            message={_t.translate("Are you sure? This action cannot be reverted.")}
            onClose={this.toggleModal}
            onConfirm={() => {
              this.toggleModal();
              this.props.onDelete(this.props.event.massage.id);
            }}
          />
        )}
      </div>
    );
  }
}

MassageEvent.propTypes = {
  /** event to be rendered */
  event: PropTypes.shape({
    bgColor: PropTypes.string,
    massage: PropTypes.shape({
      id: PropTypes.number,
      masseuse: PropTypes.shape({
        email: PropTypes.string,
        masseur: PropTypes.bool,
        name: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool,
        surname: PropTypes.string
      }),
      date: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      ending: PropTypes.oneOfType([PropTypes.number, PropTypes.instanceOf(Date)]),
      client: PropTypes.shape({
        email: PropTypes.string,
        name: PropTypes.string,
        surname: PropTypes.string,
        sub: PropTypes.string,
        subscribed: PropTypes.bool
      }),
      facility: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string
      })
    })
  }).isRequired,
  /** whether the event is to be displayed as archived or not */
  archived: PropTypes.bool,
  /** whether the event tooltip should be visible or not */
  activeTooltip: PropTypes.string,
  /** number of currently used Massage time in minutes */
  massageMinutes: PropTypes.number,
  /** function called on event assignment */
  onAssign: PropTypes.func,
  /** function called on event cancellation */
  onCancel: PropTypes.func,
  /** function called on event deletion */
  onDelete: PropTypes.func,
  /** function called on event editation */
  onEdit: PropTypes.func,
  /** function called on event more info request */
  onShowMore: PropTypes.func,
  /** current calendar view */
  view: PropTypes.string
};

MassageEvent.defaultProps = {
  archived: false,
  activeTooltip: null,
  massageMinutes: 0,
  onAssign() {},
  onCancel() {},
  onDelete() {},
  onEdit() {},
  onShowMore() {},
  view: "month"
};

export default MassageEvent;
