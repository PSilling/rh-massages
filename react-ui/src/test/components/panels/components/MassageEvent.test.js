// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import MassageEvent from "../../../../components/panels/components/MassageEvent";

test("renders content with correct props", () => {
  const testEvent = {
    massage: {
      id: 1,
      date: new Date(0),
      ending: new Date(1000),
      masseuse: {
        sub: "m-sub",
        name: "Masseuse",
        surname: "Test",
        email: "test@masseuse.org",
        subscribed: false,
        masseur: true
      },
      client: null,
      facility: { id: 1, name: "test" }
    }
  };
  const wrapper = shallow(<MassageEvent event={testEvent} view="work_week" />);
  const spans = wrapper.find("span");

  expect(spans.length).toBe(2);

  expect(wrapper).toMatchSnapshot();
});
