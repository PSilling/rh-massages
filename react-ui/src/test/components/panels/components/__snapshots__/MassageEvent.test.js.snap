// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content with correct props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MassageEvent
    activeTooltip={null}
    archived={false}
    event={
      Object {
        "massage": Object {
          "client": null,
          "date": 1970-01-01T00:00:00.000Z,
          "ending": 1970-01-01T00:00:01.000Z,
          "facility": Object {
            "id": 1,
            "name": "test",
          },
          "id": 1,
          "masseuse": Object {
            "email": "test@masseuse.org",
            "masseur": true,
            "name": "Masseuse",
            "sub": "m-sub",
            "subscribed": false,
            "surname": "Test",
          },
        },
      }
    }
    massageMinutes={Object {}}
    onAssign={[Function]}
    onCancel={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
    onShowMore={[Function]}
    view="work_week"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span>
          <span
            className="rbc-event-label"
            style={
              Object {
                "marginBottom": "3px",
                "marginTop": "2px",
              }
            }
          >
            1:00 AM — 1:00 AM
          </span>
          <span>
            <nobr>
              Masseuse
            </nobr>
          </span>
        </span>,
        false,
        false,
      ],
      "id": "Tooltip1",
      "onMouseEnter": [Function],
      "onMouseLeave": [Function],
      "style": Object {
        "minHeight": "100%",
        "minWidth": "100%",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="rbc-event-label"
              style={
                Object {
                  "marginBottom": "3px",
                  "marginTop": "2px",
                }
              }
            >
              1:00 AM — 1:00 AM
            </span>,
            <span>
              <nobr>
                Masseuse
              </nobr>
            </span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "1:00 AM — 1:00 AM",
              "className": "rbc-event-label",
              "style": Object {
                "marginBottom": "3px",
                "marginTop": "2px",
              },
            },
            "ref": null,
            "rendered": "1:00 AM — 1:00 AM",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <nobr>
                Masseuse
              </nobr>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  false,
                  "Masseuse",
                ],
              },
              "ref": null,
              "rendered": Array [
                false,
                "Masseuse",
              ],
              "type": "nobr",
            },
            "type": "span",
          },
        ],
        "type": "span",
      },
      false,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span>
            <span
              className="rbc-event-label"
              style={
                Object {
                  "marginBottom": "3px",
                  "marginTop": "2px",
                }
              }
            >
              1:00 AM — 1:00 AM
            </span>
            <span>
              <nobr>
                Masseuse
              </nobr>
            </span>
          </span>,
          false,
          false,
        ],
        "id": "Tooltip1",
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
        "style": Object {
          "minHeight": "100%",
          "minWidth": "100%",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="rbc-event-label"
                style={
                  Object {
                    "marginBottom": "3px",
                    "marginTop": "2px",
                  }
                }
              >
                1:00 AM — 1:00 AM
              </span>,
              <span>
                <nobr>
                  Masseuse
                </nobr>
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "1:00 AM — 1:00 AM",
                "className": "rbc-event-label",
                "style": Object {
                  "marginBottom": "3px",
                  "marginTop": "2px",
                },
              },
              "ref": null,
              "rendered": "1:00 AM — 1:00 AM",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <nobr>
                  Masseuse
                </nobr>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    false,
                    "Masseuse",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  false,
                  "Masseuse",
                ],
                "type": "nobr",
              },
              "type": "span",
            },
          ],
          "type": "span",
        },
        false,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
